staging/src/k8s.io/api/storage/v1beta1/types.go

```diff
@@ -209,3 +209,203 @@ type VolumeError struct {
+// kubelet plugin registration.
+// CSINodeInfo has the same name as a node. If it is missing, it means either
+// there are no CSI Drivers available on the node, or the Kubelet version is low
+// enough that it doesn't create this object.
+// CSINodeInfo has an OwnerReference that points to the corresponding node object.
+type CSINodeInfo struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// metadata.name must be the Kubernetes node name.
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// spec is the specification of CSINodeInfo
+	Spec CSINodeInfoSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
+
+	// status is the current status of CSINodeInfo
+	Status CSINodeInfoStatus `json:"status" protobuf:"bytes,3,opt,name=status"`
+}
+
+// CSINodeInfoSpec holds information about the specification of all CSI drivers installed on a node
+type CSINodeInfoSpec struct {
+	// drivers is a list of information of all CSI Drivers existing on a node.
+	// It can be empty on initialization.
+	// +patchMergeKey=name
+	// +patchStrategy=merge
+	Drivers []CSIDriverInfoSpec `json:"drivers" patchStrategy:"merge" patchMergeKey:"name" protobuf:"bytes,1,rep,name=drivers"`
+}
+
+// CSINodeInfoStatus holds information about the status of all CSI drivers installed on a node
+type CSINodeInfoStatus struct {
+	// drivers is a list of the statuses of all CSI Drivers existing on a node.
+	// It can be empty on initialization.
+	// +patchMergeKey=name
+	// +patchStrategy=merge
+	Drivers []CSIDriverInfoStatus `json:"drivers" patchStrategy:"merge" patchMergeKey:"name" protobuf:"bytes,1,rep,name=drivers"`
+}
+
+// VolumePluginMechanism is the type of mechanism components should use for volume operations
+type VolumePluginMechanism string
+
+const (
+	// VolumePluginMechanismInTree means components should use the in-tree plugin for volume operations
+	VolumePluginMechanismInTree VolumePluginMechanism = "InTree"
+	// VolumePluginMechanismCSI means components should use the CSI Driver for volume operations
+	VolumePluginMechanismCSI VolumePluginMechanism = "CSI"
+)
+
+// CSIDriverInfoStatus holds information about the status of one CSI driver installed on a node
+type CSIDriverInfoStatus struct {
+	// This is the name of the CSI driver that this object refers to.
+	// This MUST be the same name returned by the CSI GetPluginName() call for
+	// that driver. This field is required.
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+
+	// available is a boolean representing whether the driver has been installed
+	// on this node or not. It is a way for kubelet to communicate to the
+	// kube-controller-manager and scheduler about the driver's availability.
+	// This field is required. When a driver is uninstalled, available is set to false.
+	Available bool `json:"available" protobuf:"varint,2,opt,name=available"`
+
+	// volumePluginMechanism announces what mechanism underlies volume plugins.
```

Comment:
I still do not understand this field.  How is non-CSI info buried multiple levels inside an object dedicated to CSI, and keyed by a CSI driver name.  I am very confused about this.

Reply:
hope this helps? https://github.com/kubernetes/community/blob/628988e94ab3648fd784fa53318507e46d64c21d/contributors/design-proposals/storage/csi-migration.md#adc-and-kubelet-csiin-tree-sync

Reply:
@davidz627 @thockin and I discussed this.

The purpose of the volumePluginMechanism field is for kubelet to report to the AD controller whether or not csi migration has been turned on for that driver (to deal with version skew).  This is meant as a temporary field; once migration is GA and cannot be turned off, then we no longer need this bit.
  
Putting this bit in the csiNodeInfo.status makes the API lifecycle awkward: driver entries need to be added before the driver is installed so spec fields are allowed to be empty, but are actually invalid in normal operation.  The driver entry cannot be deleted even when the driver is uninstalled.  Spec and status need to be in sync.

Our proposal instead is to store this bit as an annotation in the CSINodeInfo object.  The annotation is a list of csi driver names where migration has been turned on.  Then we can get rid of the entire status subfield entirely, and restore the spec lifecycle to align with driver installation and deletion.  So the changes for this PR are:
* Remove the entire status subfield
* Add a constant to define the annotation + validation. @davidz627 to provide more details

Reply:
Do we still keep CSIDriverInfoSpec?

Reply:
Yes, let's still keep spec, so that we can make room for a status in the future (maybe for driver health monitoring?)

Reply:
Sure.